### PR TITLE
cephfs: add creation of csi default svg

### PR DESCRIPTION
**Pull request by parth-gr**
ocs operator should create the default csi svg CR

### CODE REVIEW

File: controllers/storagecluster/cephfilesystem.go
Comment (Madhu-1) [inline controllers/storagecluster/cephfilesystem.go:144]: no owner ref for this one? and also there should be a check for DeletionTimestamp to avoid updating when its under deletion

File: controllers/storagecluster/cephfilesystem.go
Comment (Madhu-1) [inline controllers/storagecluster/cephfilesystem.go:150]: SVG also need to be deleted before filesystem in `ensureDeleted` function?

File: controllers/storagecluster/cephfilesystem.go
Comment (Madhu-1) [inline controllers/storagecluster/cephfilesystem.go:135]: `createDefaultSubvolumeGroup`

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:129]: Should we track this by https://pkg.go.dev/k8s.io/api@v0.27.3/core/v1#ServiceType?

Comment (travisn) [reply controllers/storagecluster/cephfilesystem.go:129]: Why doesn't provider mode need the subvolume group?

Comment (Madhu-1) [reply controllers/storagecluster/cephfilesystem.go:129]: provider mode will not use the default csi subvolumegroup, we create an unique subvolume groups in provider as per the consumer requirements

Comment (travisn) [reply controllers/storagecluster/cephfilesystem.go:129]: From this [rook upstream discussion](https://github.com/rook/rook/issues/13040#issuecomment-1812741374), I understand Rook can always create the default subvolumegroup, so it doesn't hurt to always create this CR here even for provider mode, right?

File: controllers/storagecluster/cephfilesystem.go
Comment (Nikhil-Ladha) [reply controllers/storagecluster/cephfilesystem.go:144]: I think it might make sense to move this CR creation to a separate set of `ensureCreated` and `ensureDeleted` functions under a new file names `cephsubvolumegroup.go` and then call it in the reconcile loop accordingly.
cc @malayparida2000 @iamniting

File: controllers/storagecluster/cephfilesystem.go
Comment (malayparida2000) [inline controllers/storagecluster/cephfilesystem.go:136]: Does the name need to be fixed as "csi"? If not we should be generated the name from the filesystem name by appending, subvolumegroup at the end.

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:144]: Actually it need all the filesystem name, So we need to query each thing again. And it should be created only after the filesystem get created succefully

File: controllers/storagecluster/cephfilesystem.go
Comment (Madhu-1) [reply controllers/storagecluster/cephfilesystem.go:129]: Yes, it doesn't hurt but it's an extra resource that will not be used.

File: controllers/storagecluster/cephfilesystem.go
Comment (Nikhil-Ladha) [reply controllers/storagecluster/cephfilesystem.go:144]: Yeah, regarding that.
Looking at the code, it doesn't seem like we are creating multiple cephfilsystems, or are we?

Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:144]: For internal mode we may do

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:136]: It would be fixed at is default

File: controllers/storagecluster/cephfilesystem.go
Comment (Madhu-1) [reply controllers/storagecluster/cephfilesystem.go:144]: we create only one filesystem as of today

Comment (Nikhil-Ladha) [reply controllers/storagecluster/cephfilesystem.go:144]: Right, so we should consider that as well while implementing this CR creation.

File: controllers/storagecluster/cephfilesystem.go
Comment (Madhu-1) [reply controllers/storagecluster/cephfilesystem.go:136]: we dont have any problem this is today as we have single filesystem and we need the SVG name with `csi`

**Comment by parth-gr** [comment]
Blocked by https://github.com/rook/rook/issues/13220

**Comment by Madhu-1** [comment]
> Blocked by [rook/rook#13220](https://github.com/rook/rook/issues/13220)

@parth-gr why this blocker? please check my comment https://github.com/red-hat-storage/ocs-operator/pull/2271#discussion_r1395895269

**Comment by parth-gr** [comment]
> > Blocked by [rook/rook#13220](https://github.com/rook/rook/issues/13220)
> 
> @parth-gr why this blocker? please check my comment [#2271 (comment)](https://github.com/red-hat-storage/ocs-operator/pull/2271#discussion_r1395895269)

Okay Thanks, will follow up with this PR

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go]: Will update it added a TODO once the rook bug is fixed

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:171]: Added

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [inline controllers/storagecluster/cephfilesystem.go:127]: @Madhu-1 please confirm this

File: controllers/storagecluster/cephfilesystem.go
Comment (Nikhil-Ladha) [inline controllers/storagecluster/cephfilesystem.go:149]: We shouldn't log as `Uninstall: ...`, we are in the creation loop, if the CR doesn't exist then we simply log `Creating ...` and create the CR in the next line.
Check the above code of `cephFileSystem` creation for ref.

File: controllers/storagecluster/cephfilesystem.go
Comment (Nikhil-Ladha) [inline controllers/storagecluster/cephfilesystem.go:236]: SVG needs to be deleted before cephFileSystem, right?
So, we should move this to line:224, i.e, before the Deletion of the filesystem.

**Review by Nikhil-Ladha — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go]: IMO it wont be a problem but still moved to avoid any race condition

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go]: Updated, thanks for the feedback

**Review by BlaineEXE — COMMENTED** [review]
This looks good to me. 

The only thing I don't see is a unit test. Does the ocs-operator project expect PRs like this to have unit tests?

**Comment by malayparida2000** [comment]
> This looks good to me.
> 
> The only thing I don't see is a unit test. Does the ocs-operator project expect PRs like this to have unit tests?

@parth-gr Let's add some here to check the new CR creation

File: controllers/storagecluster/cephfilesystem.go
Comment (iamniting) [reply controllers/storagecluster/cephfilesystem.go:127]: @nb-ohad Can you pls confirm?

File: controllers/storagecluster/cephfilesystem.go
Comment (iamniting) [inline controllers/storagecluster/cephfilesystem.go:219]: Can you add a log line here?

File: controllers/storagecluster/cephfilesystem.go
Comment (iamniting) [inline controllers/storagecluster/cephfilesystem.go:146]: Can we pls rename `initialization` in this log?

File: controllers/storagecluster/cephfilesystem.go
Comment (iamniting) [inline controllers/storagecluster/cephfilesystem.go:160]:
```suggestion
		r.Log.Error(err, "Could not create/update default csi cephFilesystemSubVolumeGroup.", "cephFilesystemSubVolumeGroup", klog.KRef(cephFilesystemSubVolumeGroup.Namespace, cephFilesystemSubVolumeGroup.Name))
```

File: controllers/storagecluster/cephfilesystem.go
Comment (iamniting) [inline controllers/storagecluster/cephfilesystem.go:142]: Can we name this SVG as a filesystem name instead of a generic name `csi`? Also lets create a constant for this as this name is used at multiple places.

**Review by iamniting — CHANGES_REQUESTED** [review]

**Comment by parth-gr** [comment]
> > This looks good to me.
> > The only thing I don't see is a unit test. Does the ocs-operator project expect PRs like this to have unit tests?
> 
> @parth-gr Let's add some here to check the new CR creation

Added

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go]: Updated

File: controllers/storagecluster/cephfilesystem.go
Comment (parth-gr) [inline controllers/storagecluster/cephfilesystem.go:180]: @iamniting Can you confirm the timestamp part
I used a similar one of the filesystem.

when it sets zero `.IsZero()`

Comment (iamniting) [reply controllers/storagecluster/cephfilesystem.go:180]: IsZero to true means it is nil and not marked for deletion.

File: controllers/storagecluster/cephfilesystem_test.go
Comment (iamniting) [inline controllers/storagecluster/cephfilesystem_test.go:73]: Can you use the same constant here for the Name?

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/cephfilesystem.go
Comment (nb-ohad) [reply controllers/storagecluster/cephfilesystem.go:129]: If the intention here is to skip the default subvolumegropup for provider mode then it seems correct as we create a subvolumegropup per storageclassclaim

Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:129]: Thanks

Comment (parth-gr) [reply controllers/storagecluster/cephfilesystem.go:129]: Yes you thats correct we need to skip it for provider mode

**Comment by parth-gr** [comment]
Testing wise we have some changes in csi too, so I think we can just merge it and let build verify it.

I am not sure on creating a setup where csi and ocs operator are using a new image

**Comment by agarwal-mudit** [comment]
Can we merge this? what's pending? @iamniting

**Comment by Nikhil-Ladha** [comment]
> Testing wise we have some changes in csi too, so I think we can just merge it and let build verify it.
> 
> I am not sure on creating a setup where csi and ocs operator are using a new image

We can do a test build using this branch and the main branch of ceph-csi to see if everything works in that case, we don't need to merge and then rollback/fix later on if something isn't working.

**Comment by agarwal-mudit** [comment]
> > Testing wise we have some changes in csi too, so I think we can just merge it and let build verify it.
> > I am not sure on creating a setup where csi and ocs operator are using a new image
> 
> We can do a test build using this branch and the main branch of ceph-csi to see if everything works in that case, we don't need to merge and then rollback/fix later on if something isn't working.

We can merge it in main itself instead of creating a test build.

**Review by iamniting — APPROVED** [review]

**Pull request merged**
